include/soci/rowset.h

```diff
@@ -41,10 +41,6 @@ class rowset_iterator
     rowset_iterator(statement & st, T & define)
         : st_(&st), define_(&define)
     {
-        assert(0 != st_);
```

Comment:
I have 2 problems with using asserts: first one is the clear potential for the ODR violations when they're used in headers. Granted, usually this will be harmless, but sometimes it can result in really weird and hard to debug problems, so, after having encountered such problems one time too many, UI now try to avoid any possibility of ODR violations in the first place.

The second objection is more fundamental and is that different people use asserts for different things and you never know what they're for. Your use of them for documentation purposes only is one common case but another common case is for enforcing the preconditions and this is very different. Personally I think they are inappropriate for both of these purposes: when you see a pointer being dereferenced in the code, it's a clear documentation of the fact that it's not supposed to be used already and, anyhow, documentation shouldn't be executable code. As for preconditions, they should be handled more explicitly and, notably, always and not just when `NDEBUG` happens to be undefined.